std-bits/table/src/main/java/org/enso/table/data/column/operation/StorageIterators.java

```diff
@@ -487,4 +582,238 @@ public static <T> ColumnStorage<T> mapOverBooleanStorage(
+      }
+
+      context.safepoint();
+    }
+
+    return builder.seal();
+  }
+
+  /**
+   * Zips two long storages together, applying an operation to each pair of values. The operation's
+   * result is appended to the builder. The builderConstructor will be passed the expected size to
+   * create a new builder. If skipNothing is true, then if either value is Nothing, the result will
+   * be Nothing and appended automatically.
+   *
+   * @param source1 the first source storage to read from and iterate over.
+   * @param source2 the second source storage to read from and iterate over.
+   * @param builderConstructor a function to create a new builder of the correct type.
+   * @param skipNothing if true, then if either value is Nothing, the result will be Nothing.
+   * @param operation a callback to process a pair of values.
+   * @return a built ColumnStorage from sealing the builder.
+   * @param <T> Output Java type for the storage.
+   */
+  public static <T> ColumnStorage<T> zipOverLongStorages(
+      ColumnLongStorage source1,
+      ColumnLongStorage source2,
+      LongFunction<BuilderForType<T>> builderConstructor,
+      boolean skipNothing,
+      LongZipOperation<T> operation) {
+    if (source1 instanceof ColumnLongStorageWithArray longArrayStorage1
+        && source2 instanceof ColumnLongStorageWithArray longArrayStorage2) {
+      return zipOverLongArrayStorages(
+          longArrayStorage1, longArrayStorage2, builderConstructor, skipNothing, operation);
+    }
+
+    long size1 = source1.getSize();
+    long size2 = source2.getSize();
+
+    long size = Math.max(size1, size2);
+    var builder = builderConstructor.apply(size);
+
+    Context context = Context.getCurrent();
+
+    for (long index = 0; index < size; index++) {
+      boolean isNothing1 = index >= size1 || source1.isNothing(index);
+      boolean isNothing2 = index >= size2 || source2.isNothing(index);
+      if (skipNothing && (isNothing1 || isNothing2)) {
+        builder.appendNulls(1);
+      } else {
+        long value1 = isNothing1 ? 0 : source1.getItemAsLong(index);
+        long value2 = isNothing2 ? 0 : source2.getItemAsLong(index);
+        var result = operation.apply(index, value1, isNothing1, value2, isNothing2);
+        builder.append(result);
+      }
+
+      context.safepoint();
+    }
+
+    return builder.seal();
+  }
+
+  private static <T> ColumnStorage<T> zipOverLongArrayStorages(
+      ColumnLongStorageWithArray source1,
+      ColumnLongStorageWithArray source2,
+      LongFunction<BuilderForType<T>> builderConstructor,
+      boolean skipNothing,
+      LongZipOperation<T> operation) {
+    var data1 = source1.getArray();
+    long size1 = source1.getSize();
+    var data2 = source2.getArray();
+    long size2 = source2.getSize();
+
+    long size = Math.max(size1, size2);
+    var builder = builderConstructor.apply(size);
+
+    Context context = Context.getCurrent();
+
+    for (int index = 0; index < size; index++) {
```

Comment:
Again, given that `index` is `int` and `size` is `long`, an assertion would be good before the loop that checks if the loop even has a chance to ever terminate.